sensor_msgs_py/sensor_msgs_py/point_cloud2.py

```diff
@@ -67,56 +71,60 @@ def read_points(cloud, field_names=None, skip_nans=False, uvs=[]):
     :param skip_nans: If True, then don't return any point with a NaN value.
                       (Type: Bool, Default: False)
     :param uvs: If specified, then only return the points at the given
-        coordinates. (Type: Iterable, Default: empty list)
-    :return: Generator which yields a list of values for each point.
+        coordinates. (Type: Iterable, Default: None)
+    :return: Numpy array containing all points.
     """
     assert isinstance(cloud, PointCloud2), \
         'cloud is not a sensor_msgs.msg.PointCloud2'
-    fmt = _get_struct_fmt(cloud.is_bigendian, cloud.fields, field_names)
-    width, height, point_step, row_step, data, isnan = \
-        cloud.width, cloud.height, \
-        cloud.point_step, cloud.row_step, \
-        cloud.data, math.isnan
 
-    unpack_from = struct.Struct(fmt).unpack_from
+    assert all(x.datatype==cloud.fields[0].datatype for x in cloud.fields), \
+        'cloud fields have non equal data types'
+
+    # Create a list containing the indices of all used fields
+    if field_names is None:
+        field_ids = list(range(len(cloud.fields)))
+    else:
+        field_ids = list(map(
+            lambda field: field[0],
+            filter(
+                lambda field: field[1].name in field_names,
+                enumerate(cloud.fields))))
+
+    # Cast bytes to numpy array
+    points = np.ndarray(
+        shape=(cloud.width, len(cloud.fields)),
+        # Limitation: only one datatype per point cloud is possible, not per field
```

Comment:
Numpy arrays are only able to contain one datatype. A change that would resolve this issue but breaks the current API would be to return an array for each "collum"/field. This could be implemented by parsing the bytes first as an uint8 or similar so each byte gets one slot in the array. The array would be reshaped so each point gets its own row of bytes. The rows are then casted by the byte size of the fields into a new array with the desired field data type. This is a bit slower than the currently proposed approach, but it should still be significantly faster than the existing implantation.
In addition to that a new API function could be added to query a user defined number of uniform fields as a single array.

Reply:
Regarding breaking API changes, did this package already work in ros2 due to the PyKDL issue. If not I would be less concerned regarding breaking the API, because people coming from ros1 already face a significant number of breaking changes.

Reply:
Maybe line 80 could be changed to an `if` to use the fast implementation if possible and resort to the old implementation otherwise

Reply:
Oh sorry I confused this PR with the one in `tf2_sensor_msgs`. It indeed does not depend on PyKDL.

I also missed that combinded dtype arrays are indeed possible in numpy as long as the field length stays the same in all over all elements. So I give it a try.

I think NumPy arrays are Iterable so returning a numpy array should comply with the documented API.

Keeping both implemetations would also be an option but I would prefere to have two functions insead of an `if` in this case.

Reply:
I adapted the code to support differently typed fields.